core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -1136,7 +1145,45 @@ wasmtime_ssp_fd_renumber(
     fd_object_release(fo);
     --ft->used;
 
+    // Handle renumbering of any preopened resources
+    struct fd_prestat *prestat_from;
+    __wasi_errno_t prestat_from_error =
+        fd_prestats_get_entry(prestats, from, &prestat_from);
+
+    struct fd_prestat *prestat_to;
+    __wasi_errno_t prestat_to_error =
+        fd_prestats_get_entry(prestats, to, &prestat_to);
+
+    // Renumbering over two preopened resources.
+    if (prestat_from_error == 0 && prestat_to_error == 0) {
+        error = fd_prestats_remove_entry(prestats, to);
+
+        if (error == 0) {
+            prestat_to->dir = bh_strdup(prestat_from->dir);
+
+            if (prestat_to->dir != NULL) {
+                error = fd_prestats_remove_entry(prestats, from);
+            }
+        }
+    }
+    // Renumbering from a non-preopened fd to a preopened fd. In this case, we
+    // can't a keep the destination fd entry in the preopened table so remove
+    // it entirely.
+    else if (prestat_from_error != 0 && prestat_to_error == 0) {
+        error = fd_prestats_remove_entry(prestats, to);
+    }
+    // Renumbering from a preopened fd to a non-preopened fd
+    else if (prestat_from_error == 0 && prestat_to_error != 0) {
+        error = fd_prestats_remove_entry(prestats, from);
```

Comment:
not sure why would we remove the `from` entry here? I'd assume that both file descriptors should remain valid.